File: src/samples/python/standalone.py
# Copyright (c) 2018-2020 Scott McMillan <scott.andrew.mcmillan@gmail.com>
#
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.

# This sample is a complete standalone program to check that 1) at least 5%
# of the "/" file system is free and 2) the amount of physical memory
# is 4GB.  Data building blocks are called to collect the raw data,
# and the raw data is verified using check building blocks.
#
# The sample also demonstrates how to use results.  The individual
# results are children of the top level result.

from __future__ import print_function

import json
import logging
import sys

try:
    import wassail
except:
    # If the wassail module is installed in a non-default location,
    # you may need to set the PYTHONPATH environment variable to that
    # location.  Also use the same python version the module was built
    # for, typically the latest version of python present on the
    # system.
    print('Unable to load the wassail module')
    exit(1)

def str_issue(issue):
    """Return a string corresponding the result issue enum"""

    if issue == wassail.issue_t.YES:
        return 'YES'
    elif issue == wassail.issue_t.NO:
        return 'NO'
    elif issue == wassail.issue_t.MAYBE:
        return 'MAYBE'
    else:
        return 'UNKNOWN'

def str_priority(priority):
    """Return a string corresponding the result priority enum"""

    if priority == wassail.priority_t.EMERGENCY:
        return 'EMERGENCY'
    elif priority == wassail.priority_t.ALERT:
        return 'ALERT'
    elif priority == wassail.priority_t.ERROR:
        return 'ERROR'
    elif priority == wassail.priority_t.WARNING:
        return 'WARNING'
    elif priority == wassail.priority_t.NOTICE:
        return 'NOTICE'
    elif priority == wassail.priority_t.INFO:
        return 'INFO'
    elif priority == wassail.priority_t.DEBUG:
        return 'DEBUG'
    else:
        return 'UNKNOWN'

def print_result(result, level=0):
    """Print a wassail result"""

    print(' ' * level, 'brief: {0} {1} {2}'.format(
            result.brief, str_priority(result.priority),
            str_issue(result.issue)))
    if result.detail:
        print(' ' * (level + 2), 'detail: {}'.format(result.detail))

    for child in result.children:
        print_result(child, level + 2)

if __name__ == '__main__':
    overall_result = wassail.result('Standalone wassail sample')

    # checks
    disk = wassail.check.disk.percent_free('/', 5.0)
    memory = wassail.check.memory.physical_size(4 * 1024*1024*1024, 1 * 1024)

    # data sources
    getfsstat = wassail.data.getfsstat()
    getmntent = wassail.data.getmntent()
    sysconf = wassail.data.sysconf()
    sysctl = wassail.data.sysctl()
    sysinfo = wassail.data.sysinfo()

    # perform disk check
    if getfsstat.enabled():
      r = disk.check(getfsstat)
      overall_result.add_child(r)
    elif getmntent.enabled():
      r = disk.check(getmntent)
      overall_result.add_child(r)
    else:
      print("Unable to perform disk check - no data")

    # perform memory check
    if sysconf.enabled():
      r = memory.check(sysconf)
      overall_result.add_child(r)
    elif sysinfo.enabled():
      r = memory.check(sysinfo)
      overall_result.add_child(r)
    elif sysctl.enabled():
      r = memory.check(sysctl)
      overall_result.add_child(r)
    else:
      print("Unable to perform memory check - no data")

    # update the overall result based on the children
    overall_result.issue = overall_result.max_issue()
    overall_result.priority = overall_result.max_priority()

    # print the result
    print_result(overall_result)

File: src/samples/c++/standalone.cpp
/* Copyright (c) 2018-2020 Scott McMillan <scott.andrew.mcmillan@gmail.com>
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

/* This sample is a complete standalone program to check that 1) at least 5%
 * of the "/" file system is free and 2) the amount of physical memory
 * is 4GB.  Data building blocks are called to collect the raw data,
 * and the raw data is verified using check building blocks.
 *
 * The sample also demonstrates how to use results.  The individual
 * results are children of the top level result.
 */

#include <cstdlib>
#include <iostream>
#include <unistd.h>
#include <wassail/wassail.hpp>

int main(int argc, char **argv) {
  wassail::initialize();

  /* overall result */
  auto overall_result = wassail::make_result();
  overall_result->brief = "Standalone wassail sample";

  /* checks */
  auto disk = wassail::check::disk::percent_free("/", 5.0);
  auto memory =
      wassail::check::memory::physical_size(4L * 1024 * 1024 * 1024, 1L * 1024);

  /* data sources */
  auto getfsstat = wassail::data::getfsstat();
  auto getmntent = wassail::data::getmntent();
  auto sysconf = wassail::data::sysconf();
  auto sysctl = wassail::data::sysctl();
  auto sysinfo = wassail::data::sysinfo();

  /* perform disk check */
  if (getfsstat.enabled()) {
    auto r = disk.check(getfsstat);
    overall_result->add_child(r);
  }
  else if (getmntent.enabled()) {
    auto r = disk.check(getmntent);
    overall_result->add_child(r);
  }
  else {
    std::cerr << "Unable to perform disk check - no data" << std::endl;
  }

  /* perform memory check */
  if (sysconf.enabled()) {
    auto r = memory.check(sysconf);
    overall_result->add_child(r);
  }
  else if (sysinfo.enabled()) {
    auto r = memory.check(sysinfo);
    overall_result->add_child(r);
  }
  else if (sysctl.enabled()) {
    auto r = memory.check(sysctl);
    overall_result->add_child(r);
  }
  else {
    std::cerr << "Unable to perform memory check - no data" << std::endl;
  }

  /* update the overall result based on the children */
  overall_result->priority = overall_result->max_priority();
  overall_result->issue = overall_result->max_issue();

  /* print the result */
  std::cout << overall_result;

  return 0;
}
